WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt

```diff
@@ -0,0 +1,187 @@
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.wordpress.android.R
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.modules.UI_THREAD
+import org.wordpress.android.ui.pages.SnackbarMessageHolder
+import org.wordpress.android.ui.posts.ProgressDialogUiState
+import org.wordpress.android.ui.posts.ProgressDialogUiState.HiddenProgressDialog
+import org.wordpress.android.ui.posts.ProgressDialogUiState.VisibleProgressDialog
+import org.wordpress.android.ui.posts.editor.EditorMedia
+import org.wordpress.android.ui.posts.editor.EditorMediaListener
+import org.wordpress.android.ui.posts.editor.EditorTracker
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingMediaIdle
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingMultipleMedia
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingSingleMedia
+import org.wordpress.android.ui.utils.UiString.UiStringRes
+import org.wordpress.android.util.FluxCUtilsWrapper
+import org.wordpress.android.util.MediaUtilsWrapper
+import org.wordpress.android.util.helpers.MediaFile
+import org.wordpress.android.viewmodel.SingleLiveEvent
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Processes a list of media in the background (optimizing, resizing, rotating, etc.) and adds them to
+ * the editor one at a time.
+ */
+class OptimizeAndAddMediaToEditorUseCase @Inject constructor(
+    private val editorTracker: EditorTracker,
+    private val mediaUtilsWrapper: MediaUtilsWrapper,
+    private val fluxCUtilsWrapper: FluxCUtilsWrapper,
+    @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : CoroutineScope {
+    private var job: Job = Job()
+
+    override val coroutineContext: CoroutineContext
+        get() = mainDispatcher + job
+
+    private val _uiState: MutableLiveData<AddMediaToEditorUiState> = MutableLiveData()
+    val uiState: LiveData<AddMediaToEditorUiState> = _uiState
+
+    private val _snackBarMessage = SingleLiveEvent<SnackbarMessageHolder>()
+    val snackBarMessage = _snackBarMessage as LiveData<SnackbarMessageHolder>
+
+    init {
+        _uiState.value = AddingMediaIdle
+    }
+
+    fun optimizeAndAddAsync(
+        uriList: List<Uri>,
+        site: SiteModel,
+        isNew: Boolean,
+        editorMediaListener: EditorMediaListener,
+        editorMedia: EditorMedia
+    ) {
+        launch {
+            _uiState.value = if (uriList.size > 1) {
```

Comment:
Should this be a `postValue` call instead since it's in a bg thread?

Reply:
Default thread is actually mainThread - see https://github.com/wordpress-mobile/WordPress-Android/blob/dcd4896da50dc64764136b564dacced69d4fe9f3/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt#L50.


Most of the logic still runs on the UIThread because most of the utility methods access view - eg. toast/snackbar. However, the most heavy work - image optimization - works on a background thread.

Reply:
🤦‍♂